File: sub-system/src/main/java/org/github/controller/ErrorController.java
package org.github.controller;

import org.springframework.boot.autoconfigure.web.ServerProperties;
import org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController;
import org.springframework.boot.autoconfigure.web.servlet.error.ErrorViewResolver;
import org.springframework.boot.web.servlet.error.ErrorAttributes;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;

import javax.servlet.http.HttpServletRequest;
import java.util.List;
import java.util.Map;

import static cn.hutool.core.bean.BeanUtil.beanToMap;
import static org.github.spring.restful.json.JSONReturn.warn;
import static org.springframework.http.ResponseEntity.badRequest;

@Controller
@RequestMapping("${server.error.path:${error.path:/error}}")
public class ErrorController extends BasicErrorController {
    public ErrorController(ErrorAttributes errorAttributes, ServerProperties serverProperties, List<ErrorViewResolver> errorViewResolvers) {
        super(errorAttributes, serverProperties.getError(), errorViewResolvers);
    }

    @RequestMapping
    @Override
    public ResponseEntity<Map<String, Object>> error(HttpServletRequest request) {
        return badRequest().body(beanToMap(warn()));
    }
}

File: sub-system/src/main/java/org/github/controller/ErrorController.kt
package org.github.controller

import org.apache.shiro.ShiroException
import org.github.exception.ParamsErrorException
import org.github.exception.RemoteErrorException
import org.github.ops.error
import org.github.ops.hasChinese
import org.github.ops.log
import org.github.spring.restful.json.JSONReturn.*
import org.springframework.http.HttpStatus.resolve
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class ErrorController {
  private val log = ErrorController::class.log

  @ExceptionHandler(Throwable::class)
  fun handleThrowable(e: Throwable) = error().apply {if (e.localizedMessage.hasChinese()) withRetMsg(e.localizedMessage)}.let {ResponseEntity(it, requireNotNull(resolve(it.retCode)))}.also {log.error(e) {}}

  @ExceptionHandler(ParamsErrorException::class)
  fun handleParamsErrorException(e: ParamsErrorException) = warn().withRetMsg(e.localizedMessage).let {ResponseEntity(it, requireNotNull(resolve(it.retCode)))}.also {log.error(e) {}}

  @ExceptionHandler(RemoteErrorException::class)
  fun handleRemoteErrorException(e: RemoteErrorException) = e.data.let {ResponseEntity(it, requireNotNull(resolve(it.retCode)))}.also {log.error(e) {}}

  @ExceptionHandler(ShiroException::class)
  fun handleShiroException(e: ShiroException) = auth().apply {withRetMsg(if (e.localizedMessage.hasChinese()) e.localizedMessage else "权限不足")}.let {ResponseEntity(it, requireNotNull(resolve(it.retCode)))}.also {log.error(e) {}}
}
